File: Calculo Numerico/Trabalho 15/Trabalho15.py
import math


def print_res(res):
	for x, y in res:
		print("y(%.2f) = %.4f" % (x, y))


def eq(x, y):
	return 2*x*y

def next_y(x, y, h):


	incog = [0 for i in range(4)]

	incog[0] = h*eq(x, y)

	x_aux = x + 0.5*h
	y_aux = y + 0.5*incog[0]
	incog[1] = h*eq(x_aux, y_aux)

	x_aux = x + 0.5*h
	y_aux = y + 0.5*incog[1]
	incog[2] = h*eq(x_aux, y_aux)

	x_aux = x + h
	y_aux = y + incog[2]
	incog[3] = h*eq(x_aux, y_aux)

	total = incog[0]+2*incog[1]+2*incog[2]+incog[3]
	return y + (1/6.0)*total


def runge_kutta(x0, xn, y0, h):

	y = []
	y.append((x0, y0))

	x = x0+h
	while x <= xn+h:
		y.append((x, next_y(y[-1][0], y[-1][1], h)))
		x += h 

	return y


result = runge_kutta(1, 1.5, 1, 0.1)
print_res(result)

File: Calculo Numerico/Trabalho 15/Trabalho15.cpp
/*
* Trabalho 15.
* Implemente o método de Runge-Kutta de 4a Ordem
*
* Calculo Numerico
* Claudio Schepke
* ---
* Ciencia da Computacao
* Universidade Federal do Pampa
* 
*
* Integrantes:
* ---
* Marcos V. Treviso, 121150107, marcosvtreviso@gmail.com
*
*
* Instrucoes:
* ---
* Compile: g++ -o trab Trabalho15.cpp -Wall -Wextra
* Run: ./trab
*
*/

#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <cstring>
#include <vector>
#include <limits>

using namespace std;

const double EPSILON = 1e-7;

typedef pair<double, double> dd;
typedef vector<dd> vd;



void printResult(vd &v){
	int n = v.size();
	printf("Xn:    Yn:\n");
	for(int i=0; i<n; i++)
		printf("%.2f   %.4f\n", v[i].first, v[i].second);
}


double eq(double x, double y){
	return 2*x*y;
}

double nextY(double h, double x0, double y0){

	double incog[4];
	double x_aux, y_aux, total;

	incog[0] = h*eq(x0, y0);

	x_aux = x0 + 0.5*h;
	y_aux = y0 + 0.5*incog[0];
	incog[1] = h*eq(x_aux, y_aux);

	x_aux = x0 + 0.5*h;
	y_aux = y0 + 0.5*incog[1];
	incog[2] = h*eq(x_aux, y_aux);

	x_aux = x0 + h;
	y_aux = y0 + incog[2];
	incog[3] = h*eq(x_aux, y_aux);

	total = incog[0]+2*incog[1]+2*incog[2]+incog[3];
	return y0 + (1/6.0)*total;
}


vd rungeKutta(double x0, double xn, double y0, double h){

	vd y;
	int m;

	y.push_back(dd(x0, y0));

	for(double x=x0+h; x<xn+EPSILON; x+=h){
		m = y.size()-1;
		y.push_back(dd(x, nextY(h, y[m].first, y[m].second)));
	}

	return y;
}



int main(){

	vd result = rungeKutta(1, 1.5, 1, 0.1);
	printResult(result);

	return 0;
}